exec/java-exec/src/main/java/org/apache/drill/exec/metastore/analyze/AnalyzeInfoProviderRegistry.java

```diff
@@ -0,0 +1,128 @@
+import org.apache.drill.common.scanner.persistence.ScanResult;
+import org.apache.drill.exec.physical.base.GroupScan;
+import org.apache.drill.metastore.exceptions.MetastoreException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Class is responsible for dynamic initialization of all implementations
+ * of {@link AnalyzeInfoProvider} annotated with {@link AnalyzeInfoProviderTemplate} and
+ * returning required {@link AnalyzeInfoProvider} instance for specific {@link GroupScan} implementation.
+ * <p>
+ *
+ * AnalyzeInfoProvider list initialization is delayed until {@link #get(GroupScan)} method is called.
+ * AnalyzeInfoProvider implementations must have constructor without arguments.
+ */
+public class AnalyzeInfoProviderRegistry {
+  private static final Logger logger = LoggerFactory.getLogger(AnalyzeInfoProviderRegistry.class);
+
+  private final ScanResult classpathScan;
+  private volatile List<AnalyzeInfoProvider> analyzeInfoProviders;
+
+  public AnalyzeInfoProviderRegistry(ScanResult classpathScan) {
+    this.classpathScan = classpathScan;
+  }
+
+  /**
+   * Returns {@link AnalyzeInfoProvider} instance suitable for specified {@link GroupScan}.
+   *
+   * @param groupScan group scan
+   * @return {@link AnalyzeInfoProvider} instance
+   */
+  public AnalyzeInfoProvider get(GroupScan groupScan) {
+    if (analyzeInfoProviders == null) {
+      synchronized (this) {
+        if (analyzeInfoProviders == null) {
+          analyzeInfoProviders = Collections.unmodifiableList(init());
+        }
+      }
+    }
+
+    List<AnalyzeInfoProvider> suitableProviders = analyzeInfoProviders.stream()
+        .filter(analyzeInfoProvider -> analyzeInfoProvider.supportsGroupScan(groupScan))
+        .collect(Collectors.toList());
+
+    if (suitableProviders.isEmpty()) {
+      throw new MetastoreException(String.format("Unable to find AnalyzeInfoProvider implementation for GroupScan class [%s]",
+          groupScan.getClass().getSimpleName()));
+    } else if (suitableProviders.size() > 1) {
+      logger.warn("Found several AnalyzeInfoProvider implementations for the same GroupScan class [{}}]. " +
+          "Please specify only unique AnalyzeInfoProvider implementation for every group scan using [{}] annotation " +
+              "to prevent using arbitrary implementation.",
+          groupScan.getClass().getSimpleName(), AnalyzeInfoProviderTemplate.class.getName());
+    }
+
+    return suitableProviders.get(0);
+  }
+
+  private List<AnalyzeInfoProvider> init() {
+    List<AnalyzeInfoProvider> providers;
+    List<AnnotatedClassDescriptor> annotatedClasses = classpathScan.getAnnotatedClasses(AnalyzeInfoProviderTemplate.class.getName());
```

Comment:
Why we annotate classes if we can simply look for classes which implement `AnalyzeInfoProvider` interface thus annotation can be removed.